lib/puppet/data_providers/hiera_interpolate.rb

```diff
@@ -1,161 +1,12 @@
-  end
+  include Puppet::Pops::Lookup::Interpolation
 
+  # For backward compatibility
+  # @deprecated
   def qualified_lookup(segments, value)
-    segments.each do |segment|
-      throw :no_such_key if value.nil?
-      if segment =~ /^[0-9]+$/
-        segment = segment.to_i
-        unless value.instance_of?(Array)
-          raise Puppet::DataBinding::LookupError,
-                "Data provider type mismatch: Got #{value.class.name} when Array was expected to enable lookup using key '#{segment}'"
-        end
-        throw :no_such_key unless segment < value.size
-      else
-        unless value.respond_to?(:'[]') && !(value.instance_of?(Array) || value.instance_of?(String))
-          raise Puppet::DataBinding::LookupError,
-                "Data provider type mismatch: Got #{value.class.name} when a hash-like object was expected to enable lookup using key '#{segment}'"
-        end
-        throw :no_such_key unless value.include?(segment)
-      end
-      value = value[segment]
-    end
-    value
-  end
-
-  def get_method_and_data(data, allow_methods)
-    if match = data.match(/^(\w+)\((?:["]([^"]+)["]|[']([^']+)['])\)$/)
-      raise Puppet::DataBinding::LoookupError, 'Interpolation using method syntax is not allowed in this context' unless allow_methods
-      key = match[1]
-      data = match[2] || match[3] # double or single qouted
-    else
-      key = 'scope'
-    end
-    [key, data]
-  end
-
-  # Split key into segments. A segment may be a quoted string (both single and double quotes can
-  # be used) and the segment separator is the '.' character. Whitespace will be trimmed off on
-  # both sides of each segment. Whitespace within quotes are not trimmed.
-  #
-  # If the key cannot be parsed, this method will yield a string describing the problem to a one
-  # parameter block. The block must return an exception instance.
-  #
-  # @param key [String] the string to split
-  # @return Array<String> the array of segments
-  # @yieldparam problem [String] the problem, i.e. 'Syntax error'
-  # @yieldreturn [Exception] the exception to raise
-  def split_key(key)
-    segments = key.split(/(\s*"[^"]+"\s*|\s*'[^']+'\s*|[^'".]+)/)
-    if segments.empty?
-      # Only happens if the original key was an empty string
-      ''
-    elsif segments.shift == ''
-      count = segments.size
-      raise yield('Syntax error') unless count > 0
-
-      segments.keep_if { |seg| seg != '.' }
-      raise yield('Syntax error') unless segments.size * 2 == count + 1
-      segments.map! do |segment|
-        segment.strip!
-        segment.start_with?('"') || segment.start_with?("'") ? segment[1..-2] : segment
-      end
-    else
-      raise yield('Syntax error')
-    end
+    Puppet::deprecation_warning(
+      'Puppet::DataProviders::HieraInterpolate#qualified_lookup is deprecated, use Puppet::Pops::Lookup::SubLookup#sub_lookup')
+    sub_lookup('<unknown key>', Puppet::Pops::Lookup::Invocation.new(nil), segments, value)
```

Comment:
Please note that this deprecation will only affect fairly special early adopters that a) adopted 4.x data providers, b) decided to write their own Hiera-like provider, and c), decided to reuse the functionality exposed with the #qualified_lookup function. I wouldn't be too surprised if the actual number of users that has done that is zero. Personally, I'm convinced that if any user has done that, they will appreciate the warning, especially given the context of the change because their specialized provider will henceforth not be able to a) report the full key of a subkey that misses, or b)  participate in the explanation support.